fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,479 @@
+struct ParentViewData {
+    height_data: TVar<SequentialKeyCache<BlockHeight, ParentViewPayload>>,
+}
+
+impl ParentViewData {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error> {
+        let cache = self.height_data.read()?;
+        // safe to unwrap, we dont allow no upper bound
+        Ok(cache.upper_bound())
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<BlockHash>, Error> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.0.clone()))
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.1.clone()))
+    }
+
+    fn top_down_msgs(
+        &self,
+        from_height: BlockHeight,
+        to_height: BlockHeight,
+    ) -> StmResult<Vec<CrossMsg>, Error> {
+        let cache = self.height_data.read()?;
+        let v = cache
+            .values_within(from_height, to_height)
+            .flat_map(|i| i.2.iter())
+            .cloned()
+            .collect();
+        Ok(v)
+    }
+}
+
+impl ParentViewProvider for InMemoryFinalityProvider {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error> {
+        let h = self.parent_view_data.latest_height()?;
+        Ok(h)
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<BlockHash>, Error> {
+        let v = self.parent_view_data.block_hash(height)?;
+        Ok(v)
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        let v = self.parent_view_data.validator_set(height)?;
+        Ok(v)
+    }
+
+    fn top_down_msgs(&self, height: BlockHeight) -> StmResult<Vec<CrossMsg>, Error> {
+        let v = self.parent_view_data.top_down_msgs(height, height)?;
+        Ok(v)
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmResult<(), Error> {
+        if !top_down_msgs.is_empty() {
+            // make sure incoming top down messages are ordered by nonce sequentially
+            ensure_sequential_by_nonce(&top_down_msgs)?;
+        };
+
+        let r = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache
+                .append(height, (block_hash, validator_set, top_down_msgs))
+                .map_err(Error::NonSequentialParentViewInsert);
+            (cache, r)
+        })?;
+
+        if let Err(e) = r {
+            return abort(e);
+        }
+
+        Ok(())
+    }
+}
+
+impl ParentFinalityProvider for InMemoryFinalityProvider {
+    fn last_committed_finality(&self) -> StmResult<IPCParentFinality, Error> {
+        let finality = self
+            .last_committed_finality
+            .read_clone()?
+            .ok_or(StmError::Abort(Error::CommittedFinalityNotReady))?;
```

Comment:
By the way I'm surprised that this is possible. There must always be something that we can consider final on the parent: genesis; either that of the entire parent chain, or where the subnet was conceived.

Reply:
This is what we discussed turning into a panic in the new phased initialisation: 
1. Internally the `last_committed_finality` is a `TVar<Option<...>>` and starts empty because we don't want to peak into the genesis file, and we might be restoring from a snapshot anyway that CometBFT needs to fetch for us.
2. Before the genesis is applied or the snapshot is loaded, CometBFT will not give us blocks to process, so the `App` should not call this method. 
3. If it did, that would be a programming error, and we can just `panic!` if the `last_committed_finality` is `None`. 
4. During genesis or loading a snapshot, the interpreter should call `set_committed_finality` to cause this to go become `Some`. 
5. After that it's safe to query and return `Stm<IPCParentFinality>`